subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/ArtifactViewAttributesIntegrationTest.groovy

```diff
@@ -0,0 +1,535 @@
+        run ':help'
+    }
+
+    def "test adding an attribute lately without using beforeLocking or the java plugin still produces same files"() {
+        file("producer/build.gradle").text = """
+            def attr = Attribute.of("test-attr", String)
+
+            configurations {
+                producerConf1 {
+                    attributes.attribute(attr, 'foo')
+                    outgoing.artifact file('file1')
+                }
+                producerConf2 {
+                    attributes.attribute(attr, 'bar')
+                    outgoing.artifact file('file2')
+                }
+            }
+        """
+
+        buildFile.text = """
+            def attr = Attribute.of("test-attr", String)
+
+            configurations {
+                filesConsumerConf {
+                    attributes.attribute(attr, 'foo')
+                }
+
+                artifactsConsumerConf {
+                    attributes.attribute(attr, 'foo')
+                }
+            }
+
+            dependencies {
+                filesConsumerConf(project(":producer"))
+                artifactsConsumerConf(project(":producer"))
+            }
+
+            tasks.register("verifySameFiles") {
+                def incomingFiles = configurations.filesConsumerConf.incoming.files
+                def artifactViewFiles = configurations.filesConsumerConf.incoming.artifactView {}.files
+                configurations.filesConsumerConf.attributes.attribute(attr, 'bar')
+
+                doLast {
+                    assert incomingFiles*.name == ['file2']
+                    assert artifactViewFiles*.name == ['file2']
+                }
+            }
+
+            tasks.register("verifySameArtifacts") {
+                def incomingArtifacts = configurations.artifactsConsumerConf.incoming.artifacts
+                def artifactViewArtifacts = configurations.artifactsConsumerConf.incoming.artifactView {}.artifacts
+                configurations.artifactsConsumerConf.attributes.attribute(attr, 'bar')
+
+                doLast {
+                    assert incomingArtifacts*.id.file.name == ['file2']
+                    assert artifactViewArtifacts*.id.file.name == ['file2']
+                }
+            }
+        """
+
+        expect:
+        run ':verifySameFiles'
+
+        and:
+        run ':verifySameArtifacts'
+    }
+
+    @ToBeFixedForConfigurationCache(because = "Legacy behavior is not supported with the configuration cache")
+    def "test legacy behavior when adding an attribute lately without using beforeLocking or the java plugin still produces same files"() {
+        System.setProperty(DefaultConfiguration.USE_LEGACY_ATTRIBUTE_SNAPSHOT_BEHAVIOR, Boolean.TRUE.toString())
```

Comment:
Does this work with forking tests? Traditionally for integration tests I think we supply system properties using a `-D` argument. I did a quick search and didn't find other integ tests which set the property in the test like this

Reply:
I haven't seen forking test errors, but I'll update to be consistent.